transport/src/main/java/io/netty/channel/ChannelOption.java

```diff
@@ -85,10 +88,44 @@
             valueOf("DATAGRAM_CHANNEL_ACTIVE_ON_REGISTRATION");
 
     /**
-     * Creates a new {@link ChannelOption} with the specified {@code name}.
+     * Creates a new {@link ChannelOption} with the specified {@param name} or return the already existing
+     * {@link ChannelOption} for the given name.
      */
+    @SuppressWarnings("unchecked")
     public static <T> ChannelOption<T> valueOf(String name) {
-        return new ChannelOption<T>(name);
+        checkNotNull(name, "name");
+        ChannelOption<T> option = names.get(name);
+        if (option == null) {
+            option = new ChannelOption<T>(name);
+            ChannelOption<T> old = names.putIfAbsent(name, option);
+            if (old != null) {
+                option = old;
+            }
+        }
+        return option;
+    }
+
+    /**
+     * Returns {@code true} if a {@link ChannelOption} exists for the given {@code name}.
+     */
+    public static boolean exists(String name) {
+        checkNotNull(name, "name");
+        return names.containsKey(name);
+    }
+
+    /**
+     * Creates a new {@link ChannelOption} for the given {@param name} or fail with an
+     * {@link IllegalArgumentException} if a {@link ChannelOption} for the given {@param name} exists.
+     */
+    @SuppressWarnings("unchecked")
+    public static <T> ChannelOption<T> createOrFail(String name) {
```

Comment:
newInstance() - just like we don't name pipeline.addLast() as
addLastOrFail()

--
https://twitter.com/trustin https://twitter.com/trustin_ko
https://twitter.com/netty_project


On Wed, Feb 18, 2015, at 03:41 AM, Norman Maurer wrote:
> In transport/src/main/java/io/netty/channel/ChannelOption.java[1]:


> > +    }
> +
> +    /**
> +     * Returns {@code true} if a {@link ChannelOption} exists for the
>       given {@code name}.
> +     */
> +    public static boolean exists(String name) {
> +        checkNotNull(name, "name");
> +        return names.containsKey(name);
> +    } +
> +    /**
> +     * Creates a new {@link ChannelOption} for the given {@param
>       name} or fail with an
> +     * {@link IllegalArgumentException} if a {@link ChannelOption}
>       for the given {@param name} exists.
> +     */
> +    @SuppressWarnings("unchecked")
> +    public static <T> ChannelOption<T> createOrFail(String name) {
>      you would like newInstance or newInstanceOrFail ?


> — Reply to this email directly or view it on GitHub[2].

>



Links:

  1. https://github.com/netty/netty/pull/3421#discussion_r24840981
  2. https://github.com/netty/netty/pull/3421/files#r24840981